operator/pkg/reconciliation/decommission_node.go

```diff
@@ -0,0 +1,327 @@
+	api "github.com/datastax/cass-operator/operator/pkg/apis/cassandra/v1beta1"
+	"github.com/datastax/cass-operator/operator/pkg/events"
+	"github.com/datastax/cass-operator/operator/pkg/httphelper"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+func (rc *ReconciliationContext) DecommissionNodes(epData httphelper.CassMetadataEndpoints) result.ReconcileResult {
+	logger := rc.ReqLogger
+	logger.Info("reconcile_racks::DecommissionNodes")
+	dc := rc.Datacenter
+
+	for idx := range rc.desiredRackInformation {
+		rackInfo := rc.desiredRackInformation[idx]
+		statefulSet := rc.statefulSets[idx]
+		desiredNodeCount := int32(rackInfo.NodeCount)
+		maxReplicas := *statefulSet.Spec.Replicas
+		lastPodSuffix := stsLastPodSuffix(maxReplicas)
+
+		if maxReplicas > desiredNodeCount {
+			dcPatch := client.MergeFrom(dc.DeepCopy())
+			updated := false
+
+			updated = rc.setCondition(
+				api.NewDatacenterCondition(
+					api.DatacenterScalingDown, corev1.ConditionTrue)) || updated
+
+			if updated {
+				err := rc.Client.Status().Patch(rc.Ctx, dc, dcPatch)
+				if err != nil {
+					logger.Error(err, "error patching datacenter status for scaling down rack started")
+					return result.Error(err)
+				}
+			}
+
+			rc.ReqLogger.Info(
+				"Need to update the rack's node count",
+				"Rack", rackInfo.RackName,
+				"maxReplicas", maxReplicas,
+				"desiredSize", desiredNodeCount,
+			)
+
+			rc.Recorder.Eventf(rc.Datacenter, corev1.EventTypeNormal, events.ScalingDownRack,
+				"Scaling down rack %s", rackInfo.RackName)
+
+			if err := setOperatorProgressStatus(rc, api.ProgressUpdating); err != nil {
+				return result.Error(err)
+			}
+
+			err := rc.DecommissionNodeOnRack(rackInfo.RackName, epData, lastPodSuffix)
+			if err != nil {
+				return result.Error(err)
+			}
+
+			return result.RequeueSoon(10)
+		}
+	}
+
+	return result.Continue()
+}
+
+func (rc *ReconciliationContext) DecommissionNodeOnRack(rackName string, epData httphelper.CassMetadataEndpoints, lastPodSuffix string) error {
+	for _, pod := range rc.dcPods {
+		podRack := pod.Labels[api.RackLabel]
+		if podRack == rackName && strings.HasSuffix(pod.Name, lastPodSuffix) {
+			mgmtApiUp := isMgmtApiRunning(pod)
+			if !mgmtApiUp {
+				return fmt.Errorf("Management API is not up on node that we are trying to decommission")
+			}
+
+			if err := rc.EnsurePodsCanAbsorbDecommData(pod, epData); err != nil {
+				dcPatch := client.MergeFrom(rc.Datacenter.DeepCopy())
+				updated := rc.setCondition(api.NewDatacenterCondition(api.DatacenterScaleDownFailed,
+					corev1.ConditionTrue))
+
+				if updated {
+					patchErr := rc.Client.Status().Patch(rc.Ctx, rc.Datacenter, dcPatch)
+					if patchErr != nil {
+						rc.ReqLogger.Error(patchErr, "error patching datacenter status for scale down failed")
+						return patchErr
+					}
+				}
+				return err
+			}
+
+			if err := rc.NodeMgmtClient.CallDecommissionNodeEndpoint(pod); err != nil {
+				// TODO this returns a 500 when it works
```

Comment:
ouch. :)

Reply:
did you get to test against a fixed version of the management API?

Reply:
I'd just make a ticket for us to come back and clean this up later. Make it part of the psp epic.